File: src/main/java/de/schramm/royalbash/controller/service/gameevent/CardDrawnEvent.java
package de.schramm.royalbash.controller.service.gameevent;

import de.schramm.royalbash.controller.service.core.Game;
import lombok.Builder;
import lombok.Value;

@Value
@Builder
public class CardDrawnEvent implements GameEvent {

    private final String playerId;
    private final int amountOfCards;

    @Override
    public Game invoke(Game game) {
        return game.findPlayer(playerId)
                .map(player -> game.updatePlayer(player, player.drawCards(amountOfCards)))
                .orElse(game);
    }
}

File: src/main/kotlin/de/schramm/royalbash/application/gameevent/CardDrawnEvent.kt
package de.schramm.royalbash.application.gameevent

import de.schramm.royalbash.domain.Game

data class CardDrawnEvent(
        var playerId: String,
        var amountOfCards: Int
) : GameEvent {

    constructor(): this(playerId = "", amountOfCards = 0)

    override fun invoke(game: Game): Game {
        return game.findPlayer(playerId)
                ?.let { game.updatePlayer(it, it.drawCards(amountOfCards)) }
                ?: game
    }
}
